src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,461 @@
+  // If the adjoint is assigned to a DataflowVar (the adjoint corresponds to a non-output binding),
+  // it would be stored in adjoint_var_map_ for future lookup
+  Var EmitAdjoint(const Var& source_var, const AdjointMsg& adjoint, bool is_dataflow_var) {
+    Var adjoint_var;
+    if (is_dataflow_var) {
+      adjoint_var = builder_->Emit(AdjointMsgToExpr(adjoint), source_var->name_hint() + "_adjoint");
+      adjoint_var_map_.Set(source_var, adjoint_var);
+    } else {
+      adjoint_var =
+          builder_->EmitOutput(AdjointMsgToExpr(adjoint), source_var->name_hint() + "_adjoint");
+    }
+    return adjoint_var;
+  }
+
+  // Handle the return value of the AD function.
+  // Returns the new return value, which would be like:
+  // Tuple(original_return_value,
+  //       Tuple(adjoint_of_require_grads_1, adjoint_of_require_grads_2, ...))
+  Expr Epilogue(const Array<Var>& require_grads, const Var& target_var) {
+    // create adjoint variables for inputs, and then bind adjoints
+    Array<Expr> out_adjoints;
+
+    for (Var var : require_grads) {
+      // If the var don't have adjoint msg, it do not contribute to the target
+      // so its adjoint is zeros
+      AdjointMsg adjoint =
+          adjoint_msg_map_.Get(var).value_or(InitZerosAdjointNested(GetStructInfo(var)));
+      Var adjoint_var = EmitAdjoint(var, adjoint, false);
+      out_adjoints.push_back(adjoint_var);
+    }
+
+    return Tuple(Array<Expr>{target_var, Tuple(out_adjoints)});
+  }
+
+  // The block builder of the corresponding GradientMutator, to emit bindings
+  BlockBuilder builder_;
+
+  // Forward Var to its adjoint Var
+  Map<Var, Var> adjoint_var_map_;
+  // Forward Var to its adjoint NestedMsg<Expr>
+  // We use NestedMsg<Expr> to save the adjoint information (equaivalent to adjoint Expr)
+  // When emiting, adjoint information will be transformed into adjoint Expr
+  Map<Var, AdjointMsg> adjoint_msg_map_;
+};
+
+class GradientMutator : public ExprMutator {
+ public:
+  static IRModule Transform(IRModule mod, GlobalVar gvar, Array<Var> require_grads) {
+    Function old_func = Downcast<Function>(mod->Lookup(gvar));
+    CheckRequireGrads(require_grads, old_func->params, gvar->name_hint);
+
+    Function new_func = CopyWithNewParams(old_func);
+    // map the parameter list into new params
+    for (size_t i = 0; i < require_grads.size(); ++i) {
+      int idx = std::find(old_func->params.begin(), old_func->params.end(), require_grads[i]) -
+                old_func->params.begin();
+      require_grads.Set(i, new_func->params[idx]);
+    }
+
+    GradientMutator mutator(mod, require_grads);
+    Function new_func_transformed = Downcast<Function>(mutator.VisitExpr(new_func));
+
+    IRModule new_module = GetRef<IRModule>(mod.CopyOnWrite());
+    new_module->Add(GlobalVar(gvar->name_hint + "_adjoint"), new_func_transformed);
+    return new_module;
+  }
+
+  Expr VisitExpr_(const FunctionNode* func) override {
```

Comment:
Use “final” for all the overridden methods.
```suggestion
  Expr VisitExpr_(const FunctionNode* func) final {
```